wiki/Client Security.md

```diff
@@ -0,0 +1,95 @@
+> 
+> https://github.com/polymorpher/one-wallet/blob/ae1b113cdfe022d2ec65152e3b8654f27f756d99/code/lib/onewallet.js#L25
+> 
+> Since the OTP is confined to a 6-digit numerical number, and the generation process relies on SHA256, the right OTP (and by extension, its EOTP) at any given time can be easily enumerated by brute-force, for any given leaf hash value. See for the code of doing so at 
+> 
+> https://github.com/polymorpher/one-wallet/blob/afe39e657df1522ee8941e386025c4ddafc3ab94/code/lib/onewallet.js#L141 
+
+An adequate solution must make it exceedingly difficult, if not practically impossible, for any party to perform any transfer (or similar operations) without the input from the Google Authenticator. At the same time, it must not significantly slow down the wallet creation process, or the process of making an authorized operation with the correct authenticator code.
+
+This leaves us several options to enhance the security:
+
+1. Substantially increase the search space required for brute force attack, from the current 1 million possible OTPs.
+2. Substantially increase the memory required for enumerating or caching all possibilities during a brute force attack.
+3. Substantially increase the computation cost and complexity for hash functions used in generating EOTP (Encoded OTPs) such that they are difficult to be executed in computing processors (e.g. GPU, FPGA, ASIC)
+
+### Notations
+
+We denote the size of the search space of OTPs by **M**, the number of leaves in the Merkle Tree as **N** (such that each operation requires a unique EOTP, which corresponds to each leaf). Note that **N** is also the product of the number of time intervals in the Merkle Tree, times the number of operations the client wants to execute per time interval.
+
+In a standard setup, **M** is 10<sup>6</sup> and **N** is also roughly 10<sup>6</sup>. This two numbers should not be confused with each other. 
+
+The choice of **M** is limited by the number of digits supported by Google Authenticator. In Android, this is unfortunately fixed to 6 at this time. 
+
+The choice of **N** is set by the lifespan of the wallet and the number of operations per time interval. The time interval duration is fixed at 30 seconds at this time due to limitation of Android Google Authenticator. The default lifespan of a wallet is 1 year, which leads to 3600 * 24 / 30 * 365 = 1051200 ~= 10<sup>6</sup> leafs.
+
+We also denote the hash of the OTP Seed as *k<sub>h</sub>* and the OTP Seed itself as *k*, similar to the notation in [Protocol](https://github.com/polymorpher/one-wallet/blob/43197c70e36cb58c2884c423c1e665feff232042/wiki/protocol.pdf).  
+
+As described in Protocol § 1.1.5, each leaf node at index *i* is computed by the following process, which we will denote as ** GenLeaf**:
+
+1. Taking first 26 bytes of *k<sub>h</sub>*, and concatenate with 2 bytes of nonce, and 4 bytes of OTP corresponding to index *i*. All bytes are in big endian encoding.
```

Comment:
From @ivan-homoliak-sutd  : 

> What is nonce, and who generates and stores it? If it is stored by the client, then to me it means that it is the same as the h_k

Reply:
The nonce is a counter per time interval that is visible on the contract, to enable multiple operations per time interval. Since it is public, it needs not to be considered for the purpose of security strength analysis here. The nonce is implemented, but not documented at this time (not a priority). I will document it in the next major revision of the Wiki.

Reply:
OK, thanks.

Reply:
BTW, I do not understand how there can be **multiple valid operations** per a single time interval. Intuitively, only a single valid operation can be submitted at the time since the user can see only a single TOTP in GA. Even if there were submitted two, the reveal stage would finish only the first one.

Reply:
For each time interval, we encode `maxNumOperationsPerInterval` number of different EOTPs during setup by inserting different nonces at the input. See https://github.com/polymorpher/one-wallet/blob/1ed4417b9ba7c6d6cf57b3a0a311023fca800b9a/code/lib/onewallet.js#L25. The smart contract keeps track of the expected nonce for each time interval, and increment the nonce by one upon a success reveal. The client gets the expected nonce from the smart contract (or simply use 0, if `maxNumOperationsPerInterval=1`) and use that as part of the input in addition to the OTP for the current time interval. Since only EOTP is revealed, an attacker would not be able to know the true OTP, hence cannot use the nonce to forge an operation.

Reply:
We clarified with Aaron today, and now I agree that it makes sense principally. We also clarified that there is no 1M OTPs, but 1M times the max value of nonce (currently 256). So the client generates 256 EOTP from a single GA OTP, which is further aggregated to the Merkle tree. My suggestion is to decrease the max number of the nonce to 8 or 16, since executing more operations within the same 30s interval is unrealistic even in the case of a shared wallet.

Reply:
Correction: current max value of nonce is 1 (i.e. maxOperationsPerInterval) for all clients. The line of code referenced above says "2 bytes for nonce" only indicates that 2 bytes are reserved for nonce in the encoding procedure.

Reply:
2B => 2^16 = ~16k. Would you intend to use 1B instead? (which would enable 256 nonces/operations)

Reply:
Yes but that may make operations less efficient, since a WORD is 16-bits and usually operations are done most efficiently on 4-bytes at a time (e.g. integer size). Having 1-byte in the middle of all these misaligns the rest of the stuff and make them positioned at odd bytes